Add orientation to VMobject, shoelace, ArcPolygon, Tiling, Graph

XorUnison opened the pull request:
~Added shoelace method to space_ops.py~
~Added orientation methods to vectorized_mobject.py, uses shoelace~
Added ~ArcPolygon~, Tiling and Graph to geometry.py
Added Honeycomb to three_dimensions.py
~Added BraceBetweenPoints to brace.py~

The code I used for testing the new additions:
```python
class Test1(ZoomedScene):# Orientation
    CONFIG = {
        "zoomed_camera_frame_starting_position": [0,0,0],
        "zoomed_display_corner": [0,0,0],
        "zoomed_display_height": config['frame_height'],
        "zoomed_display_width": config['frame_width'],
        "zoom_factor": 0.5,
    }   
    def construct(self):
        self.activate_zooming(animate=False)
        pol1=RegularPolygon(5)
        pol2=RegularPolygon(5).force_direction("CW")
        pol3=RegularPolygon(5).reverse_direction()
        pol4=RegularPolygon(5).reverse_direction().reverse_direction()
        pol5=RegularPolygon(5).force_direction("CW").reverse_direction()
        self.play(ShowCreation(pol1))
        self.remove(pol1)
        self.play(ShowCreation(pol2))
        self.remove(pol2)
        self.play(ShowCreation(pol3))
        self.remove(pol3)
        self.play(ShowCreation(pol4))
        self.remove(pol4)
        self.play(ShowCreation(pol5))
        self.wait(1)
```
![Test1](https://user-images.githubusercontent.com/2881103/88482443-89450280-cf61-11ea-8197-3d7eed35de68.gif)
```python
class Test2(Scene):# ArcPolygon
    def construct(self):
        r3 = math.sqrt(3)
        arc_config = {"stroke_width":3,"stroke_color":RED,
                      "fill_opacity":0.5,"color": GREEN}
        pol_config = {"stroke_width":10,"stroke_color":BLUE,
                      "fill_opacity":1,"color": PURPLE}
        arc0=ArcBetweenPoints(np.array([-1,0,0]),
                              np.array([1,0,0]),radius=2,**arc_config)
        arc1=ArcBetweenPoints(np.array([1,0,0]),
                              np.array([0,r3,0]),radius=2,**arc_config)
        arc2=ArcBetweenPoints(np.array([0,r3,0]),
                              np.array([-1,0,0]),radius=2,**arc_config)
        reuleaux_triangle=ArcPolygon(arc0,arc1,arc2,**pol_config)
        self.play(Write(reuleaux_triangle))
        self.wait(2)
```
![Test2](https://user-images.githubusercontent.com/2881103/88482461-9bbf3c00-cf61-11ea-83d4-32a825d6ddb3.gif)
```python
class Test3(Scene):# Tiling
    def construct(self):
        test=Tiling(Square(),
                    [[Mobject.shift,[2.1,0,0]]],
                    [[Mobject.shift,[0,2.1,0]]],
                    range(-1,1),
                    range(-1,1))
        self.play(Write(test))
        self.play(FadeOut(test.tile_dictionary[1][-1]))
        self.wait(2)
```
![Test3](https://user-images.githubusercontent.com/2881103/88482464-a1b51d00-cf61-11ea-929d-af8604d2d6d9.gif)
```python
class Test4(ZoomedScene):# Graph
    CONFIG = {
        "zoomed_camera_frame_starting_position": [0,0,0],
        "zoomed_display_corner": [0,0,0],
        "zoomed_display_height": config['frame_height'],
        "zoomed_display_width": config['frame_width'],
        "zoom_factor": 0.5,
    }   
    def construct(self):
        self.activate_zooming(animate=False)
        g = {0: [[0,0,0], [1, 2], {"color": BLUE}],
             1: [[1,0,0], [0, 2], {"color": GRAY}],
             2: [[0,1,0], [0, 1], {"color": PINK}]}
        K3 = Graph(g,vertex_config={"radius": 0.2,"fill_opacity": 1},
                   edge_config={"stroke_width": 5,"color": RED})
        K3v = K3.vertices
        K3e = K3.edges
        self.play(ShowCreation(K3e),ShowCreation(K3v))
        self.wait(2)
        self.remove(K3e,K3v)
        
        g = {0: [[0,0,0], [[1,{"angle": 2}], [2,{"color": WHITE}]], {"color": BLUE}],
             1: [[1,0,0], [0, 2], {"color": GRAY}],
             2: [[0,1,0], [0, 1], {"color": PINK}]}
        K3 = Graph(g,vertex_config={"radius": 0.2,"fill_opacity": 1},
                   edge_config={"stroke_width": 5,"color": RED})
        K3v = K3.vertices
        K3e = K3.edges
        self.play(ShowCreation(K3e),ShowCreation(K3v))
        self.wait(2)
```
![Test4](https://user-images.githubusercontent.com/2881103/88482473-a974c180-cf61-11ea-994a-4fa5f808c00c.gif)
```python
class Test5(Scene):# BraceBetweeenPoints
    def construct(self):
        bbp=BraceBetweeenPoints([-1,0,0],[1,0,0])
        self.play(Write(bbp))
        self.wait(2)
```
Pretty self-explanatory.
```python
class Test6(ThreeDScene):# Honeycomb
    def construct(self):
        self.set_camera_orientation(0.6, -0.7853981, 86.6)
        self.begin_ambient_camera_rotation()
        
        test=Honeycomb(Cube(stroke_width=5,stroke_color=BLUE,
                            fill_opacity=0.3,fill_color=PURPLE),
                    [[Mobject.shift,[2.1,0,0]]],
                    [[Mobject.shift,[0,2.1,0]]],
                    [[Mobject.shift,[0,0,2.1]]],
                    range(-1,1),
                    range(-1,1),
                    range(-1,1))
        self.play(Write(test))
        self.wait(2)
        self.play(FadeOut(test.cell_dictionary[1][-1][1]))
        self.wait(2)
        self.move_camera(0.8*np.pi/2, -0.45*np.pi)
        self.wait(2)
```
![Test6](https://user-images.githubusercontent.com/2881103/89028346-6ed9a300-d32c-11ea-8b10-27e28257011a.gif)

huguesdevimeux commented:
Could you write a test of this PR ? Refer to the wiki to see the instructions. 

And just a request, could you add a visual of your changes ? like an image, or a gif. It would be easier for all of us to see and understand what you did. Thank you in advance !

XorUnison commented:
> 
> 
> Could you write a test of this PR ? Refer to the wiki to see the instructions.
> 
> And just a request, could you add a visual of your changes ? like an image, or a gif. It would be easier for all of us to see and understand what you did. Thank you in advance !

Sure I guess, but I'll have to work myself into the testing architecture. From the looks of it... I have to mostly add scenes to `test_geometry.py`. Also I just realized I wanted to add another obvious feature, BraceBetweenPoints, so there's that too.

huguesdevimeux commented:
> Sure I guess, but I'll have to work myself into the testing architecture. From the looks of it... I have to mostly add scenes to `test_geometry.py`. Also I just realized I wanted to add another obvious feature, BraceBetweenPoints, so there's that too.

Yes, in theory you only have to touch `test_geometry.py`.

XorUnison commented:
> Yes, in theory you only have to touch `test_geometry.py`.

But I'd wager tests should also test some consistency on applying colors etc... but I see nothing like this in the tests, nor in the wiki. Do we have any policy on that?

XorUnison commented:
Alright. BraceBetweenPoints has been added, tests for the 3 new geometry classes have been added, and the OP now contains gifs of the orientation methods and the 3 new geometry classes.

XorUnison commented:
> I think I'm speaking for @PgBiel here, but I think that it would be extremely useful if you could add `numpy` style docstrings to the objects and methods that you have added.
> 
> Having the detailed breakdown of instantiating each object (like you've done now) is really well done and detailed, but having docstrings for each and every method would be _extremely_ useful when using an IDE.

Noted. I'll have a look at that. Shouldn't be too hard to add considering it's not that many new objects/methods. And there's not really such a thing as too verbose comments.

XorUnison commented:
Alright I've added yet more verbose docstrings, fixed a few formatting issues I made myself and also added Honeycomb, which basically just Tiling in 3D.

There's some beef with the new formatting though, it's not giving me logs, and the local black installation was unwilling to check the files I tried to use it on, instead opting to complain about being unable to import stuff, so if there's something specific you need on the formatting someone else is gonna have to tell me what and how.

XorUnison commented:
> 
> 
> Suggestion: Maybe rename this to `GraphTheoryGraph` or `GraphGT` or something like that, because this is most certainly going to be confusing to newcomers.

I've been thinking about this but neither of these seems like great ideas either. I'll probably think about this some more since this isn't gonna get merged today anyway.

huguesdevimeux commented:
> > Suggestion: Maybe rename this to `GraphTheoryGraph` or `GraphGT` or something like that, because this is most certainly going to be confusing to newcomers.
> 
> I've been thinking about this but neither of these seem like great ideas either. I'll probably think about this some more since this isn't gonna get merged today anyway.
> 
> Edit: Maybe something like:
> `EdgeVertexGraph`?
> `VertexGraph`?

EdgeVertex Graph looks good to me.

Pinging Captain Docs @PgBiel to check the docstrings (if he has time, of course)

Aathish04 commented:
@XorUnison I ran your branch through black and there were about 10 files that needed reformatting, these were reformatted and that's what my commit is.

XorUnison commented:
> EdgeVertex Graph looks good to me.

EdgeVertexGraph seems a bit clunky, but it's definitely much clearer than Graph while being less clunky than GraphTheoryGraph. I adjusted it.
> @XorUnison I ran your branch through black and there were about 10 files that needed reformatting, these were reformatted and that's what my commit is.

Noted and appreciated, but a bunch of those files weren't even touched by me. Probably jankyness from the switch.

Although I also figured out along with this commit that Black ignores the very first standard outlined in PEP8 and pretends it doesn't exist, opening delimiter alignment, which also made Black instantly die as a serious PEP8 autoformatter in my eyes. Thanks to that it made some changes that are _technically_ still okay within PEP8, but positively hideous ones that aren't really in alignment with the underlying intent of PEP8 either. I mean ignoring the very first standard of PEP8...?
I'm not saying we should dig it up again and look for a better autoformatter but Black? Yeah... wasn't a good choice at all. We dun goofed on that one.

eulertour commented on manim/mobject/geometry.py:
```diff
@@ -812,3 +859,233 @@ class RoundedRectangle(Rectangle):
+    The values are lists with 3 elements. A list of coordinates,
+    a list of connected vertices, and a configuration dictionary.
+    The coordinates determine the position of the vertex.
+    The list of connected vertices determines between which vertices
+    edges are. For clarity it's possible to have an edge defined in both
```
This should specify that the list determines which neighbor vertices to which the vertex is connected. And from the code it seems like this can also be an int, so you should mention that here.

XorUnison replied:
It's enforced to be an int actually. I adapted the explanation above a bit.

eulertour commented on manim/mobject/three_dimensions.py:
```diff
@@ -136,3 +137,70 @@ def generate_points(self):
+
+class Honeycomb(Tiling):
+    """
+    Inherits from Tiling and works effectively the same, just adding
```
effectively the same -> in effectively the same way

eulertour commented on manim/mobject/three_dimensions.py:
```diff
@@ -136,3 +137,70 @@ def generate_points(self):
+    the third dimension.
+    To achieve this __init__ and apply_transforms are extended,
+    while tile_init_loop is overridden.
+    Since it's a 3D honeycomb also allows for tile_dictionary to be
```
Since it's a 3D -> Since it's 3D,
Also capitalize honeycomb

eulertour commented on manim/mobject/types/vectorized_mobject.py:
```diff
@@ -843,6 +844,51 @@ def get_subcurve(self, a, b):
+        """
+        return shoelace(self.get_start_anchors(), False, True)
+
+    def reverse_direction(self):
```
This can just be `self.points = self.points[::-1]`.

XorUnison replied:
Good catch, I wrote that when I had less clue about direct attribute access and didn't really connect the dots while copying it over. But it does also raise the point of get_points being the exact kind of pointless get function we should get rid of.

eulertour commented on manim/mobject/types/vectorized_mobject.py:
```diff
@@ -843,6 +844,51 @@ def get_subcurve(self, a, b):
+        target_direction : str ("CW" or "CCW")
+        """
+        if not (target_direction == "CW" or target_direction == "CCW"):
+            raise ValueError('Invalid input for force_direction | Use "CW" or "CCW"')
```
A period would be better than a | character here.

eulertour commented on manim/utils/space_ops.py:
```diff
@@ -221,3 +221,19 @@ def get_winding_number(points):
     return total_angle / TAU
+
+
+def shoelace(x_y, absoluteValue=False, orientation=True):
```
This should be split into two functions. Boolean parameters are generally bad and python variables should be in snake_case.

XorUnison replied:
The snake case is obvious.
I suppose I can split this into only a function returning the non absolute value and one that uses this function to return the orientation? Seems better indeed but I'll sleep over how to do that exactly.

XorUnison commented on manim/mobject/geometry.py:
```diff
@@ -743,6 +742,54 @@ def __init__(self, n=6, **kwargs):
+                "All ArcPolygon submobjects must be of" "type Arc/ArcBetweenPoints"
+            )
+        VMobject.__init__(self, **kwargs)
+        # Adding the arcs like this makes arcpolygon double as a group
```
Oh you mean the thing in line 778, arcpolygon? Yeah good catch, normally I pay attention to that but I must've missed it here.

XorUnison commented on manim/mobject/geometry.py:
```diff
@@ -743,6 +742,54 @@ def __init__(self, n=6, **kwargs):
+
+    ArcPolygon also doubles as a group for the input arcs.
+    That means these arcs can be drawn and seen separately.
+    If only the ArcPolygon itself is supposed to be visible,
+    make the arcs invisible (for example with "stroke_width": 0).
```
I've reworked the explanation into this:
```
    ArcPolygon doubles as a VGroup for the input arcs, so it stores
    the passed arcs like a VGroup would, but also additionally uses the
    arcs to generate new points for itself.
    This is so that the ArcPolygon can be directly manipulated while
    assuring that the stored arcs, accessible via ArcPolygon.arcs, still
    return correct values.

    Because arcs are stored like this, if only the generated ArcPolygon
    itself is supposed to be visible, the passed arcs have to be made
    invisible (for example with "stroke_width": 0).
```
Is that clear to you or do you see more room for improvement?

XorUnison commented on manim/mobject/geometry.py:
```diff
@@ -743,6 +742,54 @@ def __init__(self, n=6, **kwargs):
+    [a,b][b,c][c,a]
+    If they don't, the gaps will be filled in with straight lines.
+
+    ArcPolygon also doubles as a group for the input arcs.
```
Resolving this by referring to VGroup, because it inherits from VMobject like VGroup, and does the same thing VGroup does with some additions.

XorUnison commented on manim/mobject/geometry.py:
```diff
@@ -812,3 +859,233 @@ class RoundedRectangle(Rectangle):
+    different tiles or to simplify the following offset functions.
+    
+    Next are two nested lists that determine how the tiles are arranged.
+    The functions are typically Mobject.shift and Mobject.rotate.
+    Each list has to contain sublists with Function/Value pairs.
```
Will remove them and focus all of that on the Examples section.

XorUnison commented on manim/mobject/geometry.py:
```diff
@@ -812,3 +859,233 @@ class RoundedRectangle(Rectangle):
+    
+    Examples
+    --------
+    The nested lists determining arrangement need more explanation.
```
Agreed, it was made with the same intent as the reference to the examples and didn't get removed after I made that other reference.

leotrs commented:
What's the status of this PR? Can we get an update please? :) I believe I saw some discussion about this on Discord.

XorUnison commented:
> 
> 
> What's the status of this PR? Can we get an update please? :) I believe I saw some discussion about this on Discord.

Still working through all the requested changes. The only one that's currently left is a suggestion by EulerTour on how to handle the functions applied to the tiles. His suggestion is a bit more complicated for the simplest scenarios, but it is more flexible and pythonic in total, and also deals better with more complicated scenarios. So I aim to digest it and switch up the way Tiling works again, since it's supposed to be as good as possible upon merging after all. Once that's done I have to fix all the conflict piling up and then it'll be ready for review again.

PgBiel commented on manim/mobject/geometry.py:
```diff
@@ -743,6 +742,54 @@ def __init__(self, n=6, **kwargs):
+
+    Parameters
+    ----------
+    *arcs : Arc or ArcBetweenPoints
```
doc fixes
```suggestion
    arcs : Union[:class:`Arc`, :class:`ArcBetweenPoints`]
```

PgBiel commented on manim/mobject/geometry.py:
```diff
@@ -743,6 +742,54 @@ def __init__(self, n=6, **kwargs):
+    
+    Example
+    -------
+    ArcPolygon(arc0,arc1,arc2,arcN,**kwargs)
```
so that sphinx displays it properly
```suggestion
    ::
    
        ArcPolygon(arc0,arc1,arc2,arcN,**kwargs)
```

PgBiel commented on manim/mobject/geometry.py:
```diff
@@ -743,6 +742,54 @@ def __init__(self, n=6, **kwargs):
+    ----------
+    *arcs : Arc or ArcBetweenPoints
+    
+    Example
+    -------
```
```suggestion
    Examples
    --------
```

PgBiel commented on manim/mobject/geometry.py:
```diff
@@ -743,6 +742,54 @@ def __init__(self, n=6, **kwargs):
+    [a,b][b,c][c,a]
+    If they don't, the gaps will be filled in with straight lines.
+
+    ArcPolygon also doubles as a group for the input arcs.
+    That means these arcs can be drawn and seen separately.
+    If only the ArcPolygon itself is supposed to be visible,
+    make the arcs invisible (for example with "stroke_width": 0).
+    ArcPolygon.arcs can be used get values from these subarcs.
```
this shouldn't be under Examples but in description itself

PgBiel commented on manim/mobject/geometry.py:
```diff
@@ -743,6 +742,54 @@ def __init__(self, n=6, **kwargs):
+class ArcPolygon(VMobject):
+    """
+    The ArcPolygon is what it says, a polygon, but made from arcs.
+    More versatile than the standard Polygon.
```
just so it links to the class for quick access
```suggestion
    More versatile than the standard :class:`Polygon`.
```

PgBiel commented on manim/mobject/geometry.py:
```diff
@@ -812,3 +859,233 @@ class RoundedRectangle(Rectangle):
+    
+    Parameters
+    ----------
+    tile_prototype : Mobject or function(x,y) that returns a Mobject
```
I suppose x, y are ints?
```suggestion
    tile_prototype : Union[:class:`~.Mobject`, Callable[[:class:`int`, :class:`int`], :class:`~.Mobject`]]
```
also, maybe you should add a few short descriptions to each one

PgBiel commented on manim/mobject/geometry.py:
```diff
@@ -812,3 +859,233 @@ class RoundedRectangle(Rectangle):
+    Parameters
+    ----------
+    tile_prototype : Mobject or function(x,y) that returns a Mobject
+    x_offset : nested list of Mobject methods and values
```
well this is kinda annoying to give a type to but...
i feel like if we used tuples indicating a function and its values, it'd be easier to deal with; might be easier to use and more self-evident as well

without changing anything:
```suggestion
    x_offset : Iterable[Iterable[Union[Callable[[:class:`int`, :class:`int`], :class:`~.Mobject`]]]]
```
with tuples:
```suggestion
    x_offset : Iterable[Iterable[Tuple[Callable[[:class:`int`, :class:`int`], :class:`~.Mobject`]]]]
```

PgBiel commented on manim/mobject/geometry.py:
```diff
@@ -812,3 +859,233 @@ class RoundedRectangle(Rectangle):
+    ----------
+    tile_prototype : Mobject or function(x,y) that returns a Mobject
+    x_offset : nested list of Mobject methods and values
+    y_offset : nested list of Mobject methods and values
```
same as above
without changing anything:
```suggestion
    y_offset : Iterable[Iterable[Union[Callable[[:class:`int`, :class:`int`], :class:`~.Mobject`]]]]
```
with tuples:
```suggestion
    y_offset : Iterable[Iterable[Tuple[Callable[[:class:`int`, :class:`int`], :class:`~.Mobject`]]]]
```

PgBiel commented on manim/mobject/geometry.py:
```diff
@@ -812,3 +859,233 @@ class RoundedRectangle(Rectangle):
+    tile_prototype : Mobject or function(x,y) that returns a Mobject
+    x_offset : nested list of Mobject methods and values
+    y_offset : nested list of Mobject methods and values
+    x_range : range
+    y_range : range
```
same comment regarding short descriptions
```suggestion
    x_range : :class:`range`
    y_range : :class:`range`
```

PgBiel commented on manim/mobject/geometry.py:
```diff
@@ -812,3 +859,233 @@ class RoundedRectangle(Rectangle):
+    y_range : range
+    
+    
+    The tile prototype can be any Mobject (also groups) or a function.
+    The function format is function(x,y), returning a Mobject.
+    Using groups or functions allows the tiling to contain multiple
+    different tiles or to simplify the following offset functions.
+    
+    Next are two nested lists that determine how the tiles are arranged.
+    The functions are typically Mobject.shift and Mobject.rotate.
+    Each list has to contain sublists with Function/Value pairs.
+    More on this in the Examples section.
+
+    Last are two ranges: If both ranges are range(-1,1,1),
+    that would result in a square grid of 9 tiles.
+
+    A Tiling can be directly drawn like a VGroup.
+    Tiling.tile_dictionary[x][y] can be used to access individual tiles,
+    to color them for example.
```
I believe those last 3 lines should be moved up, to the description

PgBiel commented on manim/mobject/geometry.py:
```diff
@@ -812,3 +859,233 @@ class RoundedRectangle(Rectangle):
+    rotated by 180°. The tile at [x3,y0] would be moved to [2,0,0] and
+    still rotated by 180°.
+
+    Full example:
+    Tiling(Square(),
+           [[Mobject.shift,[2.1,0,0]]],
+           [[Mobject.shift,[0,2.1,0]]],
+           range(-1,1),
+           range(-1,1))
```
support highlighting
```suggestion
    Full example::
    
        Tiling(Square(),
               [[Mobject.shift,[2.1,0,0]]],
               [[Mobject.shift,[0,2.1,0]]],
               range(-1,1),
               range(-1,1))
```

PgBiel commented on manim/mobject/geometry.py:
```diff
@@ -812,3 +859,233 @@ class RoundedRectangle(Rectangle):
+    This would move the tile at [x1,y0] to [1,0,0] and rotates it 180°.
+    
+    When multiple sublists are passed, they are applied alternating.
+    Example for alternating shifting and rotating:
+    [[Mobject.shift,[1,0,0]],[Mobject.rotate,np.pi]]
```
support highlighting (apply these changes where applicable as well)
```suggestion
    Example for alternating shifting and rotating::
    
        [[Mobject.shift,[1,0,0]],[Mobject.rotate,np.pi]]
```

PgBiel commented on manim/mobject/geometry.py:
```diff
@@ -812,3 +859,233 @@ class RoundedRectangle(Rectangle):
+        """
+        Loops through the ranges, creates the tiles by copying the
+        prototype, adds them to self and sorts them into the dictionary.
+        Calls apply_transforms to apply passed methods.
```
just cuz link
```suggestion
        Calls :meth:`apply_transforms` to apply passed methods.
```

PgBiel commented on manim/mobject/geometry.py:
```diff
@@ -812,3 +859,233 @@ class RoundedRectangle(Rectangle):
+
+    def apply_transforms(self, x, y, tile):
+        """
+        Calls transform_tile once per dimension to position tiles.
```
```suggestion
        Calls :meth:`transform_tile` once per dimension to position tiles.
```

PgBiel commented on manim/mobject/geometry.py:
```diff
@@ -812,3 +859,233 @@ class RoundedRectangle(Rectangle):
+    Examples
+    --------
+    The nested lists determining arrangement need more explanation.
+    Example for a shift along the X-Axis, 1 in the positive direction:
+    [[Mobject.shift,[1,0,0]]]
```
```suggestion
    Example for a shift along the X-Axis, 1 in the positive direction::
    
        [[Mobject.shift,[1,0,0]]]
```

PgBiel commented on manim/mobject/geometry.py:
```diff
@@ -812,3 +859,233 @@ class RoundedRectangle(Rectangle):
+    
+    Parameters
+    ----------
+    graph : dict
```
dict of what? Dict[???, ???]

PgBiel commented on manim/mobject/geometry.py:
```diff
@@ -812,3 +859,233 @@ class RoundedRectangle(Rectangle):
+    Parameters
+    ----------
+    graph : dict
+    vertex_type : MobjectClass, optional (default: Circle)
```
```suggestion
    vertex_type : Type[:class:`~.Mobject`], optional
        (short desc?). Defaults to :class:`Circle`.
```

PgBiel commented on manim/mobject/geometry.py:
```diff
@@ -812,3 +859,233 @@ class RoundedRectangle(Rectangle):
+    ----------
+    graph : dict
+    vertex_type : MobjectClass, optional (default: Circle)
+    vertex_config : dict, optional
```
Dict of what? Dict[??key??, ??value??]

PgBiel commented on manim/mobject/geometry.py:
```diff
@@ -812,3 +859,233 @@ class RoundedRectangle(Rectangle):
+    graph : dict
+    vertex_type : MobjectClass, optional (default: Circle)
+    vertex_config : dict, optional
+    edge_type : MobjectClass, optional (default: ArcBetweenPoints)
```
```suggestion
    edge_type : Type[:class:`~.Mobject`], optional
        (short desc?). Defaults to :class:`ArcBetweenPoints`.
```

PgBiel commented on manim/mobject/geometry.py:
```diff
@@ -812,3 +859,233 @@ class RoundedRectangle(Rectangle):
+    vertex_type : MobjectClass, optional (default: Circle)
+    vertex_config : dict, optional
+    edge_type : MobjectClass, optional (default: ArcBetweenPoints)
+    edge_config : dict, optional
```
dict of what

PgBiel commented on manim/mobject/geometry.py:
```diff
@@ -812,3 +859,233 @@ class RoundedRectangle(Rectangle):
+
+    Examples
+    --------
+    Full example:
+    g = {0: [[0,0,0], [1, 2], {"color": BLUE}],
+         1: [[1,0,0], [0, 2], {"color": GRAY}],
+         2: [[0,1,0], [0, 1], {"color": PINK}]}
+    EdgeVertexGraph(g,vertex_config={"radius": 0.2,"fill_opacity": 1},
+                    edge_config={"stroke_width": 5,"color": RED})
```
highlighting
```suggestion
    Full example::
    
        g = {0: [[0,0,0], [1, 2], {"color": BLUE}],
             1: [[1,0,0], [0, 2], {"color": GRAY}],
             2: [[0,1,0], [0, 1], {"color": PINK}]}
        EdgeVertexGraph(g,vertex_config={"radius": 0.2,"fill_opacity": 1},
                        edge_config={"stroke_width": 5,"color": RED})
```

PgBiel commented on manim/mobject/geometry.py:
```diff
@@ -812,3 +859,233 @@ class RoundedRectangle(Rectangle):
+                    edge_config={"stroke_width": 5,"color": RED})
+
+    Individual config dictionaries can also be passed to edges.
+    Example:
+    g = {0: [[0,0,0], [[1,{"angle": 2}], [2,{"color": WHITE}]]...
```
```suggestion
    Example::
    
        g = {0: [[0,0,0], [[1,{"angle": 2}], [2,{"color": WHITE}]]...
```

PgBiel commented on manim/mobject/svg/brace.py:
```diff
@@ -132,3 +133,8 @@ def copy(self):
     CONFIG = {"label_constructor": TextMobject}
+
+
+class BraceBetweeenPoints(Brace):
```
docs?

PgBiel commented on manim/mobject/three_dimensions.py:
```diff
@@ -136,3 +137,70 @@ def generate_points(self):
+
+    Parameters
+    ----------
+    tile_prototype : Mobject or function(x,y,z) that returns a Mobject
+    x_offset : nested list of Mobject methods and values
+    y_offset : nested list of Mobject methods and values
+    z_offset : nested list of Mobject methods and values
+    x_range : range
+    y_range : range
+    z_range : range
```
apply my other comments from other reviews
note that, in all of them, i assumed that  x,y,z, are ints, but feel free to change that in the Callable signature

without changing things:
```suggestion
    tile_prototype : Union[:class:`~.Mobject`, Callable[[:class:`int`, :class:`int`, :class:`int`], :class:`~.Mobject`]]
    x_offset : Iterable[Iterable[Union[Callable[[:class:`int`, :class:`int`], :class:`~.Mobject`]]]]
    y_offset : Iterable[Iterable[Union[Callable[[:class:`int`, :class:`int`], :class:`~.Mobject`]]]]
    z_offset : Iterable[Iterable[Union[Callable[[:class:`int`, :class:`int`], :class:`~.Mobject`]]]]
    x_range : :class:`range`
    y_range : :class:`range`
    z_range : :class:`range`
```

If we change offsets to a list of lists of tuples, as proposed before:
```suggestion
    tile_prototype : Union[:class:`~.Mobject`, Callable[[:class:`int`, :class:`int`, :class:`int`], :class:`~.Mobject`]]
    x_offset : Iterable[Iterable[Tuple[Callable[[:class:`int`, :class:`int`], :class:`~.Mobject`]]]]
    y_offset : Iterable[Iterable[Tuple[Callable[[:class:`int`, :class:`int`], :class:`~.Mobject`]]]]
    z_offset : Iterable[Iterable[Tuple[Callable[[:class:`int`, :class:`int`], :class:`~.Mobject`]]]]
    x_range : :class:`range`
    y_range : :class:`range`
    z_range : :class:`range`
```

PgBiel commented on manim/mobject/three_dimensions.py:
```diff
@@ -136,3 +137,70 @@ def generate_points(self):
+    y_range : range
+    z_range : range
+    
+    Example
+    -------
```
```suggestion
    Examples
    --------
```

PgBiel commented on manim/mobject/three_dimensions.py:
```diff
@@ -136,3 +137,70 @@ def generate_points(self):
+    
+    Example
+    -------
+    Honeycomb(Cube(),
+              [[Mobject.shift,[2.1,0,0]]],
+              [[Mobject.shift,[0,2.1,0]]],
+              [[Mobject.shift,[0,0,2.1]]],
+              range(-1,1),
+              range(-1,1),
+              range(-1,1))
```
(Note that `::` alone does not print a `:` in docs - it just indicates highlighting. When, however, preceded by text, it prints a `:`.)
```suggestion
    ::
    
        Honeycomb(Cube(),
                  [[Mobject.shift,[2.1,0,0]]],
                  [[Mobject.shift,[0,2.1,0]]],
                  [[Mobject.shift,[0,0,2.1]]],
                  range(-1,1),
                  range(-1,1),
                  range(-1,1))
```

PgBiel commented on manim/mobject/three_dimensions.py:
```diff
@@ -136,3 +137,70 @@ def generate_points(self):
+        super().__init__(tile_prototype, x_offset, y_offset, x_range, y_range, **kwargs)
+        self.cell_dictionary = self.tile_dictionary
+
+    def tile_init_loop(self):
```
docs? (shouldn't we document every method? or are they too private?)

PgBiel commented on manim/mobject/types/vectorized_mobject.py:
```diff
@@ -843,6 +844,51 @@ def get_subcurve(self, a, b):
         vmob.pointwise_become_partial(self, a, b)
         return vmob
 
+    """
+    The next 3 methods are for getting or changing the direction of
+    points in 2D. The direction can either be "CCW", counterclockwise,
+    or "CW", clockwise. The default is counterclockwise. This also
+    determines how the objects are drawn which can by changed by
+    reverse_direction or force_direction. This can also be used to make
+    operations on objects that are dependent on direction much easier.
+    """
```
this likely won't get printed anywhere; consider either adding to each mobject's docstring or to VMobject

PgBiel commented on manim/mobject/types/vectorized_mobject.py:
```diff
@@ -843,6 +844,51 @@ def get_subcurve(self, a, b):
+
+        Returns
+        ------
+        str ("CW" or "CCW")
```
```suggestion
        :class:`str`
            Either `"CW"` or `"CCW"`.
```

PgBiel commented on manim/mobject/types/vectorized_mobject.py:
```diff
@@ -843,6 +844,51 @@ def get_subcurve(self, a, b):
+
+    def reverse_direction(self):
+        """
+        Reverts the point orientation, by inverting the point order.
```
```suggestion
        Reverts the point orientation, by inverting the point order.
        
        Returns
        --------
        :class:`VMobject`
            Returns self.
```

PgBiel commented on manim/mobject/types/vectorized_mobject.py:
```diff
@@ -843,6 +844,51 @@ def get_subcurve(self, a, b):
+
+        Parameters
+        ----------
+        target_direction : str ("CW" or "CCW")
```
```suggestion
        target_direction : :class:`str`
            Either `"CW"` or `"CCW"`.
```

PgBiel commented on manim/mobject/types/vectorized_mobject.py:
```diff
@@ -843,6 +844,51 @@ def get_subcurve(self, a, b):
+        ----------
+        target_direction : str ("CW" or "CCW")
+        """
+        if not (target_direction == "CW" or target_direction == "CCW"):
```
simpler
```suggestion
        if target_direction not in ("CW", "CCW"):
```

PgBiel commented on manim/mobject/types/vectorized_mobject.py:
```diff
@@ -843,6 +844,51 @@ def get_subcurve(self, a, b):
+        """
+        if not (target_direction == "CW" or target_direction == "CCW"):
+            raise ValueError('Invalid input for force_direction | Use "CW" or "CCW"')
+        if not (self.get_direction() == target_direction):
```
please no `not ==`

```suggestion
        if self.get_direction() != target_direction:
```

PgBiel commented on manim/utils/space_ops.py:
```diff
@@ -221,3 +221,19 @@ def get_winding_number(points):
     return total_angle / TAU
+
+
+def shoelace(x_y, absoluteValue=False, orientation=True):
```
docs, also snake case yeah as eulertour pointed out

PgBiel commented on manim/mobject/geometry.py:
```diff
@@ -812,3 +859,233 @@ class RoundedRectangle(Rectangle):
+    edge_config : dict, optional
+
+    
+    The keys for the graph have to be of type int in ascending order,
+    with each number denoting a vertex.
```
should be moved to `graph` short description

PgBiel commented on manim/mobject/geometry.py:
```diff
@@ -812,3 +859,233 @@ class RoundedRectangle(Rectangle):
+    
+    The keys for the graph have to be of type int in ascending order,
+    with each number denoting a vertex.
+    The values are lists with 3 elements. A list of coordinates,
+    a list of connected vertices, and a configuration dictionary.
+    The coordinates determine the position of the vertex.
+    The list of connected vertices determines between which vertices
+    edges are. For clarity it's possible to have an edge defined in both
+    directions, but it'll be drawn once, from lower to higher number.
+    For example if vertex 2 is connected to vertex 0, that's ignored.
+    The config dictionary is used while initializing the vertex and
+    will override values passed via vertex_config.
```
should be moved to the vertex short descs

PgBiel commented on manim/mobject/geometry.py:
```diff
@@ -812,3 +859,233 @@ class RoundedRectangle(Rectangle):
+        """
+        This method computes and applies the offsets for the tiles,
+        in the given dimension.
+        multiplies inputs, which requires arrays to be numpy arrays.
```
docs of parameters

PgBiel commented on manim/mobject/geometry.py:
```diff
@@ -812,3 +859,233 @@ class RoundedRectangle(Rectangle):
+    y_range : range
+    
+    
+    The tile prototype can be any Mobject (also groups) or a function.
+    The function format is function(x,y), returning a Mobject.
+    Using groups or functions allows the tiling to contain multiple
+    different tiles or to simplify the following offset functions.
```
should be moved to tile_prototype short desc

PgBiel commented on manim/mobject/geometry.py:
```diff
@@ -812,3 +859,233 @@ class RoundedRectangle(Rectangle):
+    Using groups or functions allows the tiling to contain multiple
+    different tiles or to simplify the following offset functions.
+    
+    Next are two nested lists that determine how the tiles are arranged.
+    The functions are typically Mobject.shift and Mobject.rotate.
+    Each list has to contain sublists with Function/Value pairs.
+    More on this in the Examples section.
```
should be moved to the offset short descs

PgBiel commented on manim/mobject/geometry.py:
```diff
@@ -812,3 +859,233 @@ class RoundedRectangle(Rectangle):
+    Each list has to contain sublists with Function/Value pairs.
+    More on this in the Examples section.
+
+    Last are two ranges: If both ranges are range(-1,1,1),
+    that would result in a square grid of 9 tiles.
```
should be moved to range short descs

PgBiel commented on manim/mobject/geometry.py:
```diff
@@ -743,6 +742,54 @@ def __init__(self, n=6, **kwargs):
+
+    def __init__(self, *arcs, **kwargs):
+        if not all(
+            [isinstance(m, Arc) or isinstance(m, ArcBetweenPoints) for m in arcs]
```
```suggestion
            isinstance(m, (Arc, ArcBetweenPoints)) for m in arcs
```

leotrs commented:
Hi @XorUnison do you have any updates on this branch? :)

leotrs commented:
There hasn't been any movement here for over two months, and by now it will be highly non-trivial to deal with the merge conflicts here. I vote we close this :(
